cluster-autoscaler/processors/provreq/injector.go

```diff
@@ -44,57 +45,91 @@ type ProvisioningRequestPodsInjector struct {
 	clock  clock.PassiveClock
 }
 
-// Process pick one ProvisioningRequest, update Accepted condition and inject pods to unscheduled pods list.
-func (p *ProvisioningRequestPodsInjector) Process(
-	_ *context.AutoscalingContext,
-	unschedulablePods []*apiv1.Pod,
+// IsAvailableForProvisioning checks if the provisioning request is the correct state for processing and provisioning has not been attempted recently.
+func (p *ProvisioningRequestPodsInjector) IsAvailableForProvisioning(pr *provreqwrapper.ProvisioningRequest) bool {
+	conditions := pr.Status.Conditions
+	if apimeta.IsStatusConditionTrue(conditions, v1.Failed) || apimeta.IsStatusConditionTrue(conditions, v1.Provisioned) {
+		return false
+	}
+	provisioned := apimeta.FindStatusCondition(conditions, v1.Provisioned)
+	if provisioned != nil {
+		if provisioned.Status == metav1.ConditionFalse && provisioned.LastTransitionTime.Add(defaultRetryTime).Before(p.clock.Now()) {
+			return true
+		}
+		return false
+	}
+	return true
+}
+
+// MarkAsAccepted marks the ProvisioningRequest as accepted.
+func (p *ProvisioningRequestPodsInjector) MarkAsAccepted(pr *provreqwrapper.ProvisioningRequest) error {
+	provreqconditions.AddOrUpdateCondition(pr, v1.Accepted, metav1.ConditionTrue, provreqconditions.AcceptedReason, provreqconditions.AcceptedMsg, metav1.NewTime(p.clock.Now()))
+	if _, err := p.client.UpdateProvisioningRequest(pr.ProvisioningRequest); err != nil {
+		klog.Errorf("failed add Accepted condition to ProvReq %s/%s, err: %v", pr.Namespace, pr.Name, err)
+		return err
+	}
+	return nil
+}
+
+// MarkAsFailed marks the ProvisioningRequest as failed.
+func (p *ProvisioningRequestPodsInjector) MarkAsFailed(pr *provreqwrapper.ProvisioningRequest, reason string) {
+	provreqconditions.AddOrUpdateCondition(pr, v1.Failed, metav1.ConditionTrue, provreqconditions.FailedToCreatePodsReason, reason, metav1.NewTime(p.clock.Now()))
+	if _, err := p.client.UpdateProvisioningRequest(pr.ProvisioningRequest); err != nil {
+		klog.Errorf("failed add Failed condition to ProvReq %s/%s, err: %v", pr.Namespace, pr.Name, err)
+	}
+}
+
+// GetPodsFromNextRequest picks one ProvisioningRequest meeting filter condition, update Accepted condition and inject pods to unscheduled pods list.
```

Comment:
nit: description isn't entirely accurate: it picks one ProvisioningRequest, updates its Accepted conditions (marks it as accepted), and returns the list of pods for this request. It doesn't inject them to unschedulable pods list, that's done in Process()

nit 2: unscheduled and unschedulable mean something different in CA and scheduler lingo

Reply:
Updated description to accurately describe function actions